### PR TITLE
Fixing API Gateway remaining tickets and bug fixes

### LINKED ISSUE #249: Support multiple API definitions for a domain, with conflict validations

### LINKED ISSUE #449: Make sure there are no usage Box<dyn Error> in this repository
Found we have a carry forward for very old code that used Box<dyn Error>

Instead of randomly fixing it in different PRs, let's have a separate ticket for this, and It wouldn't take much time hopefully

### PR DESCRIPTION
**Pull request by afsalthaj**
I thought of finishing as much API Gateway tickets remaining in the sprint as possible over the weekend. 

Fixes #449 
Fixes #249 
Fixes Function Lookup bug which I found when working with golem-timeline.

### CODE REVIEW

File: golem-worker-service-base/src/evaluator/evaluator_context.rs [hunk 1]
```diff
@@ -7,23 +7,59 @@ use crate::worker_binding::{RequestDetails, WorkerDetail};
...
+use golem_service_base::model::{ComponentMetadata, FunctionParameter, FunctionResult, WorkerId};
 use golem_wasm_rpc::TypeAnnotatedValue;
-use std::fmt::Display;
+use std::fmt::{Display, Formatter};
 
 #[derive(Clone)]
 pub struct EvaluationContext {
     pub variables: Option<TypeAnnotatedValue>,
-    pub analysed_functions: Vec<AnalysedFunction>,
+    pub functions: Vec<Function>,
```
Comment (afsalthaj) [inline golem-worker-service-base/src/evaluator/evaluator_context.rs:18]: @jdegoes  This is not fully what we discussed, but consider this as one step towards what we envisioned, while fixing a bug immediately

Comment (jdegoes) [reply golem-worker-service-base/src/evaluator/evaluator_context.rs:18]: Forward progress, I like it!

File: golem-worker-service-base/src/evaluator/evaluator_context.rs [hunk 2]
```diff
@@ -7,23 +7,59 @@ use crate::worker_binding::{RequestDetails, WorkerDetail};
...
+            None => write!(f, "{}", self.name),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Function {
+    pub fqn: FQN,
+    pub arguments: Vec<FunctionParameter>,
+    pub return_type: Vec<FunctionResult>,
```
Comment (jdegoes) [inline golem-worker-service-base/src/evaluator/evaluator_context.rs:54]: 👍

File: golem-worker-service-base/src/evaluator/evaluator_context.rs [hunk 3]
```diff
@@ -41,19 +77,27 @@ impl WorkerMetadataFetcher for NoopWorkerMetadataFetcher {
...
 impl EvaluationContext {
     pub fn empty() -> Self {
         EvaluationContext {
             variables: None,
-            analysed_functions: vec![],
+            functions: vec![],
         }
     }
 
+    pub fn find_function(&self, function: &str) -> Option<Function> {
```
Comment (jdegoes) [inline golem-worker-service-base/src/evaluator/evaluator_context.rs:96]: This is `O(n)` in number of functions, which is fine for now, but should change to `O(log n)` at some point.

File: golem-worker-service-base/src/evaluator/evaluator_context.rs [hunk 4]
```diff
@@ -83,15 +127,47 @@ impl EvaluationContext {
...
         worker_detail: &WorkerDetail,
         request: &RequestDetails,
-        functions: Vec<AnalysedFunction>,
+        component_metadata: ComponentMetadata,
     ) -> Self {
         let mut request_data = internal::request_type_annotated_value(request);
         let worker_data = create_record("worker", worker_detail.clone().to_type_annotated_value());
         let merged = request_data.merge(&worker_data);
 
+        let top_level_functions = component_metadata.functions();
```
Comment (jdegoes) [inline golem-worker-service-base/src/evaluator/evaluator_context.rs:136]: This is all a tremendous amount of work to do on each invocation. In the future we need to cache this.

**Comment by afsalthaj** [comment]
Making changes based on the new changes for unification

**Pull request merged**